File: M1/azor-chatdog-py/src/tools/clarification_tool.py
"""
Clarification Tool
Allows the model to ask for clarification when user's question is not precise enough.
"""

from cli import console
from cli.prompt import get_user_input


def ask_for_clarification(question: str) -> str:
    """
    Ask the user for clarification when the question is not precise enough.

    This tool allows the model to interactively request more details from the user
    when their initial question is ambiguous, lacks context, or needs refinement.

    Args:
        question: The clarifying question to ask the user

    Returns:
        str: The user's answer/clarification

    Example:
        Model receives: "zrób to"
        Model calls: ask_for_clarification("Co dokładnie chcesz, żebym zrobił?")
        User responds: "Napisz funkcję do sortowania listy"
        Model continues with full context
    """
    # Display the clarification request
    console.print_info("🤔 AZØR potrzebuje doprecyzowania...")
    console.print_assistant(f"\n{question}")

    # Get user's response
    user_response = get_user_input()

    if not user_response:
        return "Użytkownik nie podał odpowiedzi."

    return user_response


# Tool definition for Gemini function calling
CLARIFICATION_TOOL_DEFINITION = {
    "name": "ask_for_clarification",
    "description": """CALL THIS IMMEDIATELY when user's question is vague, unclear, or missing details.

Examples when to call:
- "zrób to" - unclear what to do
- "pomóż mi" - unclear with what
- "napisz kod" - missing language and purpose
- Question with pronouns like "to", "tamto" without context

This pauses the conversation, asks user for details, and returns their answer.""",
    "parameters": {
        "type": "object",
        "properties": {
            "question": {
                "type": "string",
                "description": "Your clarifying question in Polish. Examples: 'Co konkretnie chcesz, żebym zrobił?', 'W czym mogę Ci pomóc?', 'Jaki kod mam napisać i w jakim języku?'"
            }
        },
        "required": ["question"]
    }
}


# Map function name to actual function
CLARIFICATION_TOOLS_MAP = {
    "ask_for_clarification": ask_for_clarification
}

File: M1/azor-chatdog-js/src/tools/clarificationTool.ts
/**
 * Clarification Tool
 * Allows the model to ask for clarification when user's question is not precise enough.
 */

import { getUserInput } from '../cli/prompt.js';
import { printAssistant, printInfo } from '../cli/console.js';

/**
 * Ask the user for clarification when the question is not precise enough.
 *
 * This tool allows the model to interactively request more details from the user
 * when their initial question is ambiguous, lacks context, or needs refinement.
 *
 * @param question - The clarifying question to ask the user
 * @returns The user's answer/clarification
 *
 * @example
 * Model receives: "zrób to"
 * Model calls: askForClarification("Co dokładnie chcesz, żebym zrobił?")
 * User responds: "Napisz funkcję do sortowania listy"
 * Model continues with full context
 */
export async function askForClarification(question: string): Promise<string> {
  // Display the clarification request
  printInfo('🤔 AZØR potrzebuje doprecyzowania...');
  printAssistant(`\n${question}`);

  // Get user's response
  const userResponse = await getUserInput();

  if (!userResponse) {
    return 'Użytkownik nie podał odpowiedzi.';
  }

  return userResponse;
}

/**
 * Tool definition for Gemini function calling
 */
export interface ToolDefinition {
  name: string;
  description: string;
  parameters: {
    type: string;
    properties: Record<string, any>;
    required: string[];
  };
}

export const CLARIFICATION_TOOL_DEFINITION: ToolDefinition = {
  name: 'ask_for_clarification',
  description: `Use this tool when the user's question or request is unclear, ambiguous, or lacks necessary details.

Call this function when you need to:
- Understand vague or incomplete requests (e.g., "zrób to", "napraw błąd")
- Clarify which option the user prefers when multiple valid approaches exist
- Get specific parameters or details that are missing
- Resolve ambiguity in the user's intent

DO NOT use this for simple questions you can answer directly. Use it only when you genuinely need more information to provide a helpful response.

The user will see your clarifying question and provide an answer, which will be returned to you.`,
  parameters: {
    type: 'object',
    properties: {
      question: {
        type: 'string',
        description:
          "The clarifying question to ask the user. Be specific and helpful. Examples: 'Którą funkcję chcesz, żebym poprawił?', 'Czy mam użyć podejścia A czy B?', 'Jakie dane wejściowe powinna przyjmować ta funkcja?'",
      },
    },
    required: ['question'],
  },
};

/**
 * Map function name to actual function
 */
export const CLARIFICATION_TOOLS_MAP: Record<string, (args: any) => Promise<string>> = {
  ask_for_clarification: async (args: { question: string }) => askForClarification(args.question),
};
